cgsmiles/write_cgsmiles.py

```diff
@@ -0,0 +1,240 @@
+        else:
+            smiles += format_node(molecule, current)
+
+        # we add the bonding descriptors if there are any
+        if molecule.nodes[current].get('bonding', False):
+            smiles += format_bonding(molecule.nodes[current]['bonding'])
+
+        if current in atom_to_ring_idx:
+            # We're going to need to write a ring number
+            ring_idxs = atom_to_ring_idx[current]
+            for ring_idx in ring_idxs:
+                ring_bond = ring_idx_to_bond[ring_idx]
+                if ring_idx not in ring_idx_to_marker:
+                    marker = _get_ring_marker(ring_idx_to_marker.values())
+                    ring_idx_to_marker[ring_idx] = marker
+                    new_marker = True
+                else:
+                    marker = ring_idx_to_marker.pop(ring_idx)
+                    new_marker = False
+
+                if smiles_format and _write_edge_symbol(molecule, *ring_bond) and new_marker:
+                    order = molecule.edges[ring_bond].get('order', 1)
+                    smiles += order_to_symbol[order]
+
+                smiles += str(marker) if marker < 10 else '%{}'.format(marker)
+
+        if current in dfs_successors:
+            # Proceed to the next node in this branch
+            next_nodes = dfs_successors[current]
+            # ... and if needed, remember to return here later
+            branches.update(next_nodes[1:])
+            to_visit.extend(next_nodes)
+        elif branch_depth:
+            # We're finished with this branch.
+            smiles += ')'
+            branch_depth -= 1
+
+    smiles += ')' * branch_depth
+    return smiles
+
+def write_cgsmiles_graph(molecule):
+    """
+    Write a CGSmiles graph sans fragments at
+    different resolution.
+
+    Parameters
+    ----------
+    molecule: nx.Graph
+        a molecule where each node as a fragname attribute
+        that is used as name in the CGSmiles string.
+
+    Returns
+    -------
+    str
+        the CGSmiles string
+    """
+
+    cgsmiles_str = write_graph(molecule)
+    return "{" + cgsmiles_str + "}"
+
+def write_cgsmiles_fragments(fragment_dict, all_atom=True):
+    """
+    Write fragments of molecule graph. To identify the fragments
+    all nodes with the same `fragname` and `fragid` attributes
+    are considered as fragment. Bonding between fragments is
+    extracted from the `bonding` edge attributes.
+
+    Parameters
+    ----------
+    fragment_dict: dict[str, nx.Graph]
+        a dict of fragment graphs
+    all_atom: bool
+        write all atom SMILES if True (default) otherwise
+        write CGSmiles
+
+    Returns
+    -------
+    str
+    """
+    fragment_str = ""
+    for fragname, frag_graph in fragment_dict.items():
+        fragment_str += f"#{fragname}="
+        # format graph depending on resolution
+        fragment_str += write_graph(frag_graph, smiles_format=all_atom) + ","
+    fragment_str = "{" + fragment_str[:-1] + "}"
```

Comment:
I would make a list of fragment_str's for the separate fragments, then `','.join` them

Reply:
it doesn't matter does it?